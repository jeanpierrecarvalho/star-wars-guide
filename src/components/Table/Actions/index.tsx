import React, { memo } from "react";
import { Input } from "@material-tailwind/react";

interface IProps {
  searchTerm: string;
  handleSearchChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

const Actions: React.FC<IProps> = ({
  searchTerm,
  handleSearchChange,
}: IProps) => {
  return (
    <React.Fragment>
      <Input
        color="yellow"
        size="lg"
        label="Search..."
        onChange={handleSearchChange}
        value={searchTerm}
        onPointerEnterCapture={undefined}
        onPointerLeaveCapture={undefined}
        crossOrigin={undefined}
        className="text-white"
      />
    </React.Fragment>
  );
};

export default memo(Actions);
