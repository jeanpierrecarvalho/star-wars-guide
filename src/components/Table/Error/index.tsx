import React from "react";

interface IProps {
  error: string;
}

const Error: React.FC<IProps> = ({ error }: IProps) => {
  return (
    <React.Fragment>
      {" "}
      <p className="mt-8 text-center text-red-500">{error}</p>
    </React.Fragment>
  );
};

export default Error;
