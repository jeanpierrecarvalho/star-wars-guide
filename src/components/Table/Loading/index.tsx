import React from "react";

const Loading: React.FC = () => {
  return (
    <div
      className="z-10 absolute inset-0 flex justify-center items-center bg-black bg-opacity-60"
      role="status"
      aria-live="polite"
    >
      <span className="text-xl text-yellow-500">
        Patience you must have, my young Padawan...
      </span>
    </div>
  );
};

export default Loading;
