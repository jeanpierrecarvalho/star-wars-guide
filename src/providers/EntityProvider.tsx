import React, { useState, useCallback, useContext } from 'react'
import { EntityContext } from '@/contexts/EntityContext'
import { fetchEntity } from '@/services/star-wars-api'
import { ENTITY_TYPE } from '@/constants/entities'

interface IEntity {
  [key: string]: any
}

export const useEntityContext = () => {
  const context = useContext(EntityContext)
  if (!context) {
    throw new Error('useEntityContext must be used within an EntityProvider')
  }
  return context
}

export const EntityProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [data, setData] = useState<IEntity[]>([])
  const [page, setPage] = useState<number>(1)
  const [totalPages, setTotalPages] = useState<number>(1)
  const [searchTerm, setSearchTerm] = useState<string>('')
  const [loading, setLoading] = useState<boolean>(false)
  const [error, setError] = useState<string | null>(null)

  const loadData = async (type: string) => {
    setLoading(true)
    try {
      const response = await fetchEntity(ENTITY_TYPE[type].toFetch, page, searchTerm)
      setData(response.results)
      setTotalPages(Math.ceil(response.count / 10))
    } catch (err) {
      setError('Failed to load data. Please try again later.')
    } finally {
      setLoading(false)
    }
  }

  const handlePageChange = useCallback(
    (newPage: number) => {
      if (newPage > 0 && newPage <= totalPages) {
        setPage(newPage)
      }
    },
    [totalPages],
  )

  const handleSearchChange = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(e.target.value)
    setPage(1)
  }, [])

  return (
    <EntityContext.Provider
      value={{
        data,
        page,
        totalPages,
        searchTerm,
        loading,
        error,
        loadData,
        handlePageChange,
        handleSearchChange,
      }}
    >
      {children}
    </EntityContext.Provider>
  )
}
