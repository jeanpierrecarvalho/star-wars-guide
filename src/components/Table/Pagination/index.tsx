import { useEntityContext } from "@/providers/EntityProvider";

const Pagination: React.FC = () => {
  const { page, totalPages, handlePageChange } = useEntityContext();

  return (
    <div className="flex flex-row justify-between items-center">
      <button
        className={`${
          page === 1
            ? "bg-yellow-500 opacity-40 cursor-not-allowed"
            : "bg-yellow-500 hover:bg-yellow-300"
        } px-2 py-1 rounded-lg w-32 font-medium text-gray-900`}
        onClick={() => handlePageChange(page - 1)}
        disabled={page === 1}
        aria-label="Previous Page"
        aria-disabled={page === 1}
      >
        Previous
      </button>
      <span className="flex flex-row items-center text-center text-lg">
        Page {page} of {totalPages}
      </span>
      <button
        className={`${
          page === totalPages
            ? "bg-yellow-500 opacity-40 cursor-not-allowed"
            : "bg-yellow-500 hover:bg-yellow-300"
        } px-2 py-1 rounded-lg w-32 font-medium text-gray-900`}
        onClick={() => handlePageChange(page + 1)}
        disabled={page === totalPages}
        aria-label="Next Page"
        aria-disabled={page === totalPages}
      >
        Next
      </button>
    </div>
  );
};

export default Pagination;
