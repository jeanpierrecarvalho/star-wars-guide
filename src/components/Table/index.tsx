import React, { useEffect, memo } from "react";
import { ENTITY_TYPE } from "@/constants/entities";

import Loading from "@/components/Table/Loading";
import Error from "@/components/Table/Error";
import Actions from "@/components/Table/Actions";
import Pagination from "@/components/Table/Pagination";

import { useEntityContext } from "@/providers/EntityProvider";

interface IProps {
  type: string;
}

const Table: React.FC<IProps> = ({ type }: IProps) => {
  const context = useEntityContext();

  const { data, page, searchTerm, loading, error, loadData } = context;

  useEffect(() => {
    const handler = setTimeout(() => {
      loadData(type);
    }, 500);

    return () => {
      clearTimeout(handler);
    };
  }, [searchTerm, page]);

  return (
    <React.Fragment>
      <div className="mt-8">
        <div className="mt-12 px-6">
          <Actions />
        </div>

        <div>
          {loading && <Loading />}

          {error ? (
            <Error error={error} />
          ) : (
            <table
              className="border-collapse mt-8 w-full text-left table-auto"
              role="table"
              aria-label={`${type} data table`}
            >
              <thead>
                <tr>
                  {ENTITY_TYPE[type]?.attributes.map((attribute, index) => (
                    <th
                      key={index}
                      className="px-6"
                      role="columnheader"
                      scope="col"
                      aria-label={`${attribute} column`}
                    >
                      {attribute.charAt(0).toUpperCase() + attribute.slice(1)}
                    </th>
                  ))}
                </tr>
              </thead>
              <tbody>
                {data.map((row: any) => (
                  <tr key={row.name}>
                    {ENTITY_TYPE[type]?.attributes.map(
                      (attribute: string, index: number) => (
                        <td
                          className={`px-6 py-2 border-b ${
                            index === 0 && "w-64"
                          }`}
                          key={index}
                          role="cell"
                        >
                          {row[attribute]}
                        </td>
                      ),
                    )}
                  </tr>
                ))}
              </tbody>
            </table>
          )}
        </div>
        <div className="mt-12 px-6">
          <Pagination />
        </div>
      </div>
    </React.Fragment>
  );
};

export default memo(Table);
