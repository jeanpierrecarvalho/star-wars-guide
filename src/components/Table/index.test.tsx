import { render } from '@testing-library/react'
import { describe, it, expect } from 'vitest'

import Table from '@/components/Table'
import { EntityProvider } from '@/providers/EntityProvider'

describe('Table Component', () => {
  it('matches the snapshot', () => {
    const props = {
      type: 'characters',
    }

    const { container } = render(
      <EntityProvider>
        <Table {...props} />
      </EntityProvider>,
    )
    expect(container).toMatchSnapshot()
  })
})
