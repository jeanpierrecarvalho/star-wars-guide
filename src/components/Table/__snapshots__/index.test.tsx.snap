// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table Component > matches the snapshot 1`] = `
<div>
  <div
    class="mt-8"
  >
    <div
      class="mt-12 px-6"
    >
      <div
        class="relative w-full min-w-[200px] h-11"
      >
        <input
          aria-label="Search for entities"
          class="peer w-full h-full bg-transparent text-blue-gray-700 font-sans font-normal outline outline-0 focus:outline-0 disabled:bg-blue-gray-50 disabled:border-0 disabled:cursor-not-allowed transition-all placeholder-shown:border placeholder-shown:border-blue-gray-200 placeholder-shown:border-t-blue-gray-200 border focus:border-2 border-t-transparent focus:border-t-transparent placeholder:opacity-0 focus:placeholder:opacity-100 text-sm px-3 py-3 rounded-md border-blue-gray-200 focus:border-yellow-500 text-white"
          placeholder=" "
          type="search"
        />
        <label
          class="flex w-full h-full select-none pointer-events-none absolute left-0 font-normal !overflow-visible truncate peer-placeholder-shown:text-blue-gray-500 leading-tight peer-focus:leading-tight peer-disabled:text-transparent peer-disabled:peer-placeholder-shown:text-blue-gray-500 transition-all -top-1.5 peer-placeholder-shown:text-sm text-[11px] peer-focus:text-[11px] before:content[' '] before:block before:box-border before:w-2.5 before:h-1.5 before:mt-[6.5px] before:mr-1 peer-placeholder-shown:before:border-transparent before:rounded-tl-md before:border-t peer-focus:before:border-t-2 before:border-l peer-focus:before:border-l-2 before:pointer-events-none before:transition-all peer-disabled:before:border-transparent after:content[' '] after:block after:flex-grow after:box-border after:w-2.5 after:h-1.5 after:mt-[6.5px] after:ml-1 peer-placeholder-shown:after:border-transparent after:rounded-tr-md after:border-t peer-focus:after:border-t-2 after:border-r peer-focus:after:border-r-2 after:pointer-events-none after:transition-all peer-disabled:after:border-transparent peer-placeholder-shown:leading-[4.1] text-blue-gray-400 peer-focus:text-yellow-500 before:border-blue-gray-200 peer-focus:before:!border-yellow-500 after:border-blue-gray-200 peer-focus:after:!border-yellow-500"
        >
          Search...
           
        </label>
      </div>
    </div>
    <div>
      <table
        aria-label="characters data table"
        class="border-collapse mt-8 w-full text-left table-auto"
        role="table"
      >
        <thead>
          <tr>
            <th
              aria-label="name column"
              class="px-6"
              role="columnheader"
              scope="col"
            >
              Name
            </th>
            <th
              aria-label="height column"
              class="px-6"
              role="columnheader"
              scope="col"
            >
              Height
            </th>
            <th
              aria-label="mass column"
              class="px-6"
              role="columnheader"
              scope="col"
            >
              Mass
            </th>
            <th
              aria-label="birth_year column"
              class="px-6"
              role="columnheader"
              scope="col"
            >
              Birth_year
            </th>
            <th
              aria-label="gender column"
              class="px-6"
              role="columnheader"
              scope="col"
            >
              Gender
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
    <div
      class="mt-12 px-6"
    >
      <div
        class="flex flex-row justify-between items-center"
      >
        <button
          aria-disabled="true"
          aria-label="Previous Page"
          class="bg-yellow-500 opacity-40 cursor-not-allowed px-2 py-1 rounded-lg w-32 font-medium text-gray-900"
          disabled=""
        >
          Previous
        </button>
        <span
          class="flex flex-row items-center text-center text-lg"
        >
          Page 
          1
           of 
          1
        </span>
        <button
          aria-disabled="true"
          aria-label="Next Page"
          class="bg-yellow-500 opacity-40 cursor-not-allowed px-2 py-1 rounded-lg w-32 font-medium text-gray-900"
          disabled=""
        >
          Next
        </button>
      </div>
    </div>
  </div>
</div>
`;
